### PR TITLE
Add debug messages for DHCP state changes, Tftp progress, and PxeBc progress.

### PR DESCRIPTION
**Pull request by mikeytdisco**
## Description

Add DEBUG message to Dhcp4Dxe

- [ No] Breaking change?
  - Will this change break pre-existing builds or functionality without action being taken?

## How This Was Tested

Not tested

## Integration Instructions

To only get the DHCP debug messages, edit your platform build .dsc to add the debug flag to

Current:
NetworkPkg/Dhcp4Dxe/Dhcp4Dxe.inf

To get the new DEBUG_NET messages from Dhcp4Dxe:
  NetworkPkg/Dhcp4Dxe/Dhcp4Dxe.inf {
    <PcdsPatchableInModule>
      gEfiMdePkgTokenSpaceGuid.PcdDebugPrintErrorLevel|0x80084246
    <PcdsFixedAtBuild>
      gEfiMdePkgTokenSpaceGuid.PcdFixedDebugPrintErrorLevel|0x80084246
}

### CODE REVIEW

**Review by makubacki — APPROVED** [review]

File: NetworkPkg/Dhcp4Dxe/Dhcp4Impl.c [hunk 1]
```diff
@@ -788,11 +788,13 @@ EfiDhcp4Start (
   DhcpSb = Instance->Service;
 
   if (DhcpSb->DhcpState == Dhcp4Stopped) {
+    DEBUG ((DEBUG_NET, "%a: failed, in stopped state.\n", __FUNCTION__));
```
Comment (kuqin12) [inline NetworkPkg/Dhcp4Dxe/Dhcp4Impl.c]: layman question: should this be `DEBUG_NET|DEBUG_ERROR`?

File: NetworkPkg/Dhcp4Dxe/Dhcp4Io.c [hunk 2]
```diff
@@ -39,6 +39,7 @@ DhcpInitRequest (
     Status = DhcpSendMessage (DhcpSb, NULL, NULL, DHCP_MSG_DISCOVER, NULL);
 
     if (EFI_ERROR (Status)) {
+      DEBUG ((DEBUG_NET, "%a: SendError, new state is Dhcp4Init.\n", __FUNCTION__));
```
Comment (kuqin12) [inline NetworkPkg/Dhcp4Dxe/Dhcp4Io.c]: Print error status?

File: NetworkPkg/Dhcp4Dxe/Dhcp4Io.c [hunk 3]
```diff
@@ -47,6 +48,7 @@ DhcpInitRequest (
     Status = DhcpSendMessage (DhcpSb, NULL, NULL, DHCP_MSG_REQUEST, NULL);
 
     if (EFI_ERROR (Status)) {
+      DEBUG ((DEBUG_NET, "%a: SendError, new state is Dhcp4InitReboot.\n", __FUNCTION__));
```
Comment (kuqin12) [inline NetworkPkg/Dhcp4Dxe/Dhcp4Io.c]: Same here?

File: NetworkPkg/Dhcp4Dxe/Dhcp4Io.c [hunk 4]
```diff
@@ -1391,6 +1421,8 @@ DhcpRetransmit (
 
   ASSERT (DhcpSb->LastPacket != NULL);
 
+  DEBUG ((DEBUG_NET, "%a:\n", __FUNCTION__));
```
Comment (kuqin12) [inline NetworkPkg/Dhcp4Dxe/Dhcp4Io.c]: nit: 
```suggestion
  DEBUG ((DEBUG_NET, "%a: Entry...\n", __FUNCTION__));
```

File: NetworkPkg/Dhcp4Dxe/Dhcp4Io.c [hunk 5]
```diff
@@ -893,18 +909,21 @@ DhcpHandleReboot (
...
   if (DhcpSb->Para == NULL) {
+    DEBUG ((DEBUG_NET, "%a: Out of resources.\n", __FUNCTION__));
     Status = EFI_OUT_OF_RESOURCES;
     goto ON_EXIT;
   }
 
   DhcpSb->Selected = Packet;
   Status           = DhcpLeaseAcquired (DhcpSb);
   if (EFI_ERROR (Status)) {
+    DEBUG ((DEBUG_NET, "%a: Error from LeaseAcquired.\n", __FUNCTION__));
```
Comment (kuqin12) [inline NetworkPkg/Dhcp4Dxe/Dhcp4Io.c]:
```suggestion
    DEBUG ((DEBUG_NET, "%a: Error from LeaseAcquired - %r.\n", __FUNCTION__, Status));
```

File: NetworkPkg/Dhcp4Dxe/Dhcp4Io.c [hunk 6]
```diff
@@ -869,22 +879,28 @@ DhcpHandleReboot (
...
   //
   if (EFI_NTOHL (Head->YourAddr) != DhcpSb->ClientAddr) {
+    DEBUG ((DEBUG_NET, "%a: ACK didn't match offer.\n", __FUNCTION__));
     Status = EFI_DEVICE_ERROR;
     goto ON_EXIT;
   }
 
   Status = DhcpCallUser (DhcpSb, Dhcp4RcvdAck, Packet, NULL);
   if (EFI_ERROR (Status)) {
+    DEBUG ((DEBUG_NET, "%a: Error from CallUser.\n", __FUNCTION__));
```
Comment (kuqin12) [inline NetworkPkg/Dhcp4Dxe/Dhcp4Io.c]:
```suggestion
    DEBUG ((DEBUG_NET, "%a: Error from CallUser - %r.\n", __FUNCTION__, Status));
```

**Review by kuqin12 — APPROVED** [review]

**Review by cfernald — APPROVED** [review]

File: NetworkPkg/Dhcp4Dxe/Dhcp4Impl.c (hunk 1 above)
Comment (mikeytdisco) [reply NetworkPkg/Dhcp4Dxe/Dhcp4Impl.c]: Fixed.

File: NetworkPkg/Dhcp4Dxe/Dhcp4Io.c (hunk 2 above)
Comment (mikeytdisco) [reply NetworkPkg/Dhcp4Dxe/Dhcp4Io.c]: Fixed.

File: NetworkPkg/Dhcp4Dxe/Dhcp4Io.c (hunk 3 above)
Comment (mikeytdisco) [reply NetworkPkg/Dhcp4Dxe/Dhcp4Io.c]: Fixed.

**Pull request merged**